### DISCUSSION TITLE
Type of some unexpected value

### DISCUSSION THREAD

**Discussion by hubertp** [enso-org/enso#6278] [Work In Progress - Engine]
This subject came up as part of the [discussion on a PR](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java#L180).
It's trivial to replace that to throw a `Panic` but we seem to rely on that under some circumstances. 
I saw at least that we check
- type of `UnresolvedSymbol` 
- type of `AtomWithHole` 

What should be the type of those? `Any`, `Nothing`? We probably shouldn't throw a Panic then since we rely on it in a regular flow e.g. in [meta](https://github.com/enso-org/enso/blob/develop/test/Tests/src/Semantic/Meta_Spec.enso#L334).
/cc @JaroslavTulach @kustosz @radeusgd @Akirathan

**Comment by radeusgd** [comment]
1. It does not need to be a Panic, it may be a dataflow error. It's just weird to return an error as a value as is now.
2. Other thing is what we should put for the two mentioned objects.
  - `AtomWithHole` - shouldn't that be a normal `Atom`? Should it be distinguishable in any way?
  - `UnresolvedSymbol`, as I wrote in #6277, I would suggest to treat it the same way as we treat any lambda, like `x->x`. I.e. the `type_of` an `UnresolvedSymbol`, IMO should be a `Function`, as it essentially represents a shorthand notation for a lambda (`.foo` behaves the same as `x-> x.foo`).

As for `Meta.meata` - currently, a function is treated as a `Meta.Primitive`. I guess that's OK, but while we are discussing it, I think it would make most sense to introduce a `Meta.Function` variant.

**Reply by Akirathan** [reply]
> It's just weird to return an error as a value as is now.

It feels weird, from Java perspective, but if you try to `throw` a DataflowError, it will be propagated up to https://github.com/enso-org/enso/blob/develop/engine/runner/src/main/scala/org/enso/runner/Main.scala#L746, the execution will stop, and only "Execution finished with an error: PolyglotError" will be printed. So currently, it returns a dataflow error, and not panics.

`UnresolvedSymbol` is essentially treated as a `Function` internally in the engine, but does that really mean that we should return `Function` for typeof? Don't we want to be more specific and return `UnresolvedSymbol` type?

**Reply by radeusgd** [reply]
We can return `UnresolvedSymbol` if it is treated as a _subtype_ of `Function`. I'm fine with that.

I just want
```
case (.name) of
    _ : Function -> True
    _ -> False
```
to yield `True`.

As a user I want to **be able to match functions in my code**, and I don't want to care _how_ these functions were constructed: `(+)`, `.foo`, `_.foo`, `x-> x.foo`; and `foo` when declared as `foo x = ...` should _all_ match `Function`.

**Reply by Akirathan** [reply]
I have looked closely into the generated IR for `.foo` and it seems that we have a bug in our codegen phase, more specifically in `IrToTruffle$ExpressionProcessor.processName` (https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala#L1456), where we ignore the `isMethod` field, and we just return `UnresolvedSymbol`. I believe there should be an if branch for `isMethod`, and if the `IR.Name.Literal` is a method, some kind of `Function` should be created instead. cc @hubertp @JaroslavTulach what do you think?

**Reply by hubertp** [reply]
- re `AtomWithHole` - It's actually `HoleInAtom`, my typo, a field that should be filled. So nope, `Atom` would be the incorrect type since we don't know what kind of value will be there.
- re the case that matches unresolved symbol - yeah, that can probably be done. Such pattern matching still relies on `TypeOfNode` but we can probably add a specialization for that.
- re the bug in processor - I will take a closer look

**Reply by radeusgd** [reply]
Right, if it's a `HoleInAtom` that's a different conversation :) IMO returning an error in that case makes most sense. It's essentially querying metadata of a yet uninitialized field. That sounds like a thing that just should not be done.

**Reply by radeusgd** [reply]
My reasoning for error - I'd like `Meta.type_of x` to stay stable between invocations.

Of course, for uninitialized hole which _just will_ get initialized, something must change. But I think a change from error to a value makes sense - error before indicated that you 'asked too early'. Once initialized, the value is stable.

OTOH, a change from value to value is a bit less expected.

**Reply by JaroslavTulach** [reply]
> It's essentially querying metadata of a yet uninitialized field. 

Right.

> That sounds like a thing that just should not be done.

No, it shouldn't be done. `atom_with_hole` is a powerful API that should be used by people who know what they are doing to _temporarily_ construct an unfinished `Atom`.

> IMO returning an error in that case makes most sense. 

I'd return `Any`, but `Error` is probably also acceptable. It is a situation that shouldn't happen - I am even surprise it is checked somewhere!

**Reply by hubertp** [reply]
> I'd return Any, but Error is probably also acceptable. It is a situation that shouldn't happen - I am even surprise it is checked somewhere!

I reverted to dataflow error. Don't have strong opinions about that one so if you really want I can make it `Any`.
